### ISSUE TITLE
Open QgsMeshLayer and QgsMeshLayerProvider to python bindings

### ISSUE THREAD

**Issue by jmkerloch** [qgis/QGIS#50219]
### Feature description

We're trying to develop a QGIS plugin to display mesh from a spatialite database which store all the vertex and faces.

We also need to get the native mesh of the `QgsMeshLayer` so we can do some statistics on the layer

We have some issue to create the provider because some attributes and function are not available in python bindings:

# `QgsMesh`
We need to define vertices, edges and faces in `populateMesh`  function in our `QgsMeshDataSourceInterface` implementation.

But we can't update `QgsMesh` attributes because `SIP_SKIP`  is present in the public attributes.

# `QgsMeshLayer`
We can't access `nativeMesh` function because of `SIP_SKIP` 

# Feature wanted
Remove as many `SIP_SKIP` as possible in `QgsMeshLayer` `QgsMeshLayerProvider` to allow more interactions in python plugin.

### Additional context

I'm ready to create a Pull Request for this feature, I have already done some testing in one of mys QGIS branches.

**Comment by jmkerloch** [comment]
@vcloarec Was there a reason for these SIP_SKIP ? 

I managed to removed them in the PR mentionned above but I'm wondering if there are other things that I might miss.

**Comment by PeterPetrik** [comment]
The whole Mesh API is experimental, and where strictly not needed, we added SIP_SKIP. The reasoning is that if it is in python API, we need to maintain it/do not change the API. 

I was hoping to do some cleaning up of the API before removing SIP_SKIP.

**Comment by jmkerloch** [comment]
OK I understand.

For now we don't need to open everything to python bindings.

We need 2 things:
- be able to create a `QgsMeshLayerDataProvider` in a python plugin
    - we need to be able to populate the mesh : update `QgsMesh` for vertex/faces/segment definition
    - open some function for temporal capabilities.
- get vertex/faces/segment of a QgsMeshLayer
    - we don't need to open `nativeMesh` of `QgsMeshLayer`, we can add some getter for vertex/faces/segment, based on index.

**Comment by PeterPetrik** [comment]
If you have some file format, the best option is to add it to MDAL as a new driver https://github.com/lutraconsulting/MDAL/tree/master/mdal/frmts or adapt your software to produce one of the supported formats?

Maybe best to have a short call with us to discuss?

**Comment by jmkerloch** [comment]
@PeterPetrik  I agree it will be better to have a short call. How can we schedule a meeting ?

**Comment by PeterPetrik** [comment]
can you please write to me on peter.petrik@lutraconsulting.co.uk to arrange one? thanks!

**Comment by jmkerloch** [comment]
If someone is curious about our implementation of a  `QgsMeshLayerDataProvider`, you can check the code in THYRSIS plugin : https://gitlab.com/Oslandia/qgis/thyrsis/-/blob/feature/study_qgsmeshlayer/thyrsis/qgs_native_mesh_layer/spatialitemeshdataprovider.py

We implemented 2 types of provider:

- `ThyrsisTimeSpatialiteMeshDataProvider`, mesh dataset are related to time and read from numpy files
- `ThyrsisSpatialiteMeshDataProvider`, mesh dataset is not related to time and read from a spatialite database

In both cases, the mesh is defined in a spatialite database.

:warning: This is still a work in progress

Here are some screenshot of the render in QGIS (not different from standard QGIS `QgsMeshLayerDataProvider` but I was happy to get results :smile: ).
In this case the render could be nicer but we currently don't have an option to use a logarithmic color ramp.

![Screenshot from 2022-10-25 15-13-01](https://user-images.githubusercontent.com/53606373/197782543-902ac997-7a43-4202-8cee-e75e9f28d8e3.png)

Multiple dataset are available in case of dataset not related to time:

![Screenshot from 2022-10-25 15-18-10](https://user-images.githubusercontent.com/53606373/197783696-68add366-82b3-48b1-acd6-79c987717218.png)

Dataset altitude
![Screenshot from 2022-10-25 15-16-07](https://user-images.githubusercontent.com/53606373/197783223-f26d8f2e-c1ec-47fd-9087-561e8874677d.png)

Dataset potentiel
![Screenshot from 2022-10-25 15-17-34](https://user-images.githubusercontent.com/53606373/197783570-bf83f29b-b754-4fca-982a-5e6d0c542842.png)

**Comment by saberraz** [comment]
Why not having it simply as another format in MDAL? https://github.com/lutraconsulting/MDAL

**Comment by vpicavet** [comment]
The question here is not implementing a new format in MDAL versus implementing it elsewhere : it is about letting users the opportunity to plug their own behaviours to access the data. This is what QGIS also is about : a flexible framework to create your own applications.

As for implementing this specific format in a generic library distributed within a generic GIS tool, this has very little sense : the format is specific to one application and will not be used elsewhere. It would just add technical debt to MDAL and clutter the code for no reason. There is also no option for this application to change the format, due to other external dependencies which are totally out of scope and cannot be modified. 

We want to keep generic code into generic applications and libraries ( QGIS, MDAL), and not pollute them with specific code useless to other end-users. Keeping QGIS and its dependancies generic enough is essential to ensure maintenance, avoid dead code, allow for other users to share and mutualize code.

There are a lot of specific applications based on QGIS all over : they should be easy to develop, but should not be included into core libraries. This is exactly the point here : open the API, incorporate generic behaviour, leave the specific in the hands of the few interested people.

**Comment by saberraz** [comment]
MDAL was conceived to exactly handle these type of cases...a generic I/O library for mesh. Your plugin is still dependent on the mesh API (which depends on MDAL). So I really do no understand not having the format directly in MDAL. 

IMO, it should be done correctly and implemented as a format in MDAL, it is a format and best to be delegated to the dedicated library.

**Comment by wonder-sk** [comment]
@vpicavet the situation here is the same as with vectors and rasters - QGIS also does not support having custom data providers for rasters/vectors written in python - instead, developers can implement drivers in GDAL/OGR. Mesh layers with MDAL just follow this convention.

One option is to simply switch to one of the generic formats supported by MDAL, another option is to write a new driver in MDAL. When dealing with vector/raster data, one typically also tries to reuse existing formats rather than creating new data formats - even if that means doing some conversion from the "native" format...

**Comment by vpicavet** [comment]
I just said that the question was not to implement or not this format into MDAL, but to provide the opportunity for application developers to use the underlying library.

Let's ask @rouault if he would allow implementing in GDAL/OGR a format which is used by 10 people in the world, with its own specificities, linked to no other software nor official standard ? I guess the answer is a big "no way".

**Comment by wonder-sk** [comment]
> if he would allow implementing in GDAL/OGR a format which is used by 10 people in the world, with its own specificities, linked to no other software nor official standard ? I guess the answer is a big "no way".

Then the right solution would be to use an existing data format, supported by existing MDAL drivers - there are more than a dozen of driver implementations available already...

**Comment by vpicavet** [comment]
As I said, we have no option changing the format : it exists, it is used, it relies on other external ( very old ) software, there is no way to change that. We do not ask for this specific format to be supported by anyone else, neither in QGIS, MDAL. This burden has to be supported by the plugin developers and users, that is all. All we ask is to allow them to take that burden and work independently, so as not to make things complicated for QGIS developers.

Also, it IS actually possible to create new data providers for QGIS in Python, and this is great. The (old) Meshlayer plugin is an example of this capability : https://gitlab.com/Oslandia/qgis/meshlayer

I cannot understand why you would restrict such a use case, prohibit a legitimate user need, while the allowed required changes in QGIS are so minimal ?

**Comment by vpicavet** [comment]
Also, creating vector data providers is possible and tested : 
- https://www.itopen.it/qgis-vector-data-provider-python/
- https://github.com/qgis/QGIS/blob/2d1aa68f0d044f2aced7ebeca8d2fa6b754ac970/tests/src/python/test_provider_python.py

I  see no rationale to prevent it for mesh layers except limiting plugin developers and preventing legitimate use cases.

**Comment by vpicavet** [comment]
I do not get what "flexibility of the mesh layer for the future" means ?

As for thread safety, it will be the plugin's developer responsibility to ensure that her developments are thread-safe. This is not a matter for QGIS itself.

**Comment by wonder-sk** [comment]
> I do not get what "flexibility of the mesh layer for the future" means ?

If some parts of the API are not exposed, we can change the implementation without having to worry about breaking things. If they are exposed, that would be API break.

**Comment by vpicavet** [comment]
But this is a problem for the plugin developer, not for you. You can mark the API as experimental, and then break it, that's the deal. Having users is the best way to improve an API, get feedback, improve it iteratively and _together_. if nobody uses the API then you won't have to change it... This is absurd reasoning.

The API evolution problem, if communicated correctly, is the plugin developer's problem, it is not a QGIS developer burden.

1 and 2 are plugin's developers problems, saying "you can't touch this because you are going to hurt yourself" is infantilization and lack of trust.

To summarize : 
- in QGIS you actually can create new vector data providers in Python
- there is a legitimate use case to do the same with mesh data providers, and there is an example implementation of this use case
- there always will be specific format in the world for which we need custom code to handle
- everything in this area is marked as "experimental", so the API users are warned
- the proposed modifications in QGIS core are very small and have no impact elsewhere

I still do not get the reactions to this very small proposal, and I feel like there is a desire to block all plugin developments wanting to use this area of the code.

**Comment by vpicavet** [comment]
Maybe this is why you have this posture, I don't know but in case : be assured that if there are mesh formats which are generic enough, we would contribute it to MDAL, just as we did for GDAL/OGR many times. We do not work on our own, we play with the community and the rules, we are a fair and pure opensource player, and we want to contribute with developer's time as much as we can, on QGIS, MDAL, GDAL/OGR. More than often we fix other developer's mistakes, clean code, do chores and little-rewarding code and work.

**Comment by PeterPetrik** [comment]
Vincent, we really do not want to block any (plugin) developments and even less to take things personally. We are on the same side here...

As for the technical discussion, I agree with @wonder-sk that it is best to reuse existing MDAL drivers/formats (e.g.  have a python plugin to transfer your format to UGRID) and not expose API to python for this unstable mesh-area. On the same time, if you really would like to pursue the exposing and usage of "experimental" API regardless of our warnings, I am ok to have it included.

**Comment by rouault** [comment]
Not sure if I really need to bring my thoughts in this "hot" topic :-) But just one thing that has not been mentioned. In GDAL, there's the possibility to develop out-of-tree drivers as plugins, loaded by GDAL core, and it is certainly used by a number of people with in-house formats that don't make sense to contribute to the upstream project. Quickly looking at https://github.com/lutraconsulting/MDAL/blob/master/mdal/frmts/mdal_dynamic_driver.cpp, I guess this is possible with MDAL too. That said the ability to develop directly in Python can be useful, despite some limitations with Python code not being able to be run in parallel due to the Python global lock.

**Comment by PeterPetrik** [comment]
Yes, MDAL supports dynamic drivers, see https://www.mdal.xyz/tutorials/add_driver_tut.html#dynamic-drivers  Those are loaded on run-time and doesn't need to be part of the official MDAL repo
